KAFKA-12802 Added a file based cache for consumed remote log metadata for each partition to avoid consuming again incase of broker restarts.

satishd opened the pull request:
- Added snapshots for consumed remote log metadata for each partition to avoid consuming again in case of broker restarts. These snapshots are stored in the respective topic partition log directories.
- Added TopicBasedRemoteLogMetadataManagerRestartTest:
  * loads the earlier saved snapshots after restart
  * checks the entries are available
  * starts the consumer and add more metadata entries
  * checks the newly added entries and loaded entries are available
 
### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

satishd committed:
KAFKA-12802 Added a file based cache for consumed remote log metadata for each partition to avoid consuming again incase of broker restarts.
- Added snapshots for consumed remote log metadata for each partition to avoid consuming again incase of broker restarts. These snapshots are stored in the respective topic partition log directories.
- Added TopicBasedRemoteLogMetadataManagerRestartTest:
  * loads the earlier saved snapshots after restart
  * checks the entries are available
  * starts the consumer and add more metadata entries
  * checks the newly addded entries and loaded entries are available

satishd commented:
This is built on top of https://github.com/apache/kafka/pull/10579. We can review and merge once #10579 is merged.

satishd commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/CommittedOffsetsFile.java:
```diff
@@ -0,0 +1,86 @@
+/**
+ * This class represents a file containing the committed offsets of remote log metadata partitions.
+ */
+public class CommittedOffsetsFile {
```
This file will be enhanced to use `SnapshotFile` from #11060 once it is merged.

satishd replied:
This is updated with the latest commit.

satishd commented:
@junrao: Please review this PR once https://github.com/apache/kafka/pull/11060 is reviewed and merged.

satishd commented:
@junrao gentle reminder to review the changes.

junrao commented:
@satishd : Could you rebase this PR? Thanks.

satishd commented:
@junrao Rebased the PR and resolved the conflicts, thanks.

junrao commented on storage/src/main/resources/message/RemoteLogSegmentMetadataRecordSnapshot.json:
```diff
@@ -0,0 +1,92 @@
+  "type": "data",
+  "name": "RemoteLogSegmentMetadataRecordSnapshot",
+  "validVersions": "0",
+  "flexibleVersions": "none",
```
Should we support flexible version from the beginning so that we could potentially support downgrade during future format changes?

satishd replied:
This was an old version and merged into this branch. Corrected with the latest branch.

junrao commented on storage/src/main/resources/message/RemoteLogSegmentMetadataRecordSnapshot.json:
```diff
@@ -0,0 +1,92 @@
+      "name": "SegmentLeaderEpochs",
+      "type": "[]SegmentLeaderEpochEntry",
+      "versions": "0+",
+      "about": "Event timestamp of this segment.",
```
This description of the field seems incorrect.

junrao commented on storage/src/main/resources/message/RemoteLogSegmentMetadataRecordSnapshot.json:
```diff
@@ -0,0 +1,92 @@
+      "name": "BrokerId",
+      "type": "int32",
+      "versions": "0+",
+      "about": "Broker (controller or leader) id from which this event is created. DELETE_PARTITION_MARKED is sent by the controller. DELETE_PARTITION_STARTED and DELETE_PARTITION_FINISHED are sent by remote log metadata topic partition leader."
```
The description seems to be intended for RemotePartitionDeleteMetadataSnapshot?

junrao commented on storage/src/main/resources/message/RemotePartitionDleteMetadataSnapshot.json:
```diff
@@ -0,0 +1,42 @@
+      "name": "EventTimestamp",
+      "type": "int64",
+      "versions": "0+",
+      "about": "Event timestamp of this segment."
```
This is not about a segment.

ccding commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/CommittedOffsetsFile.java:
```diff
@@ -0,0 +1,80 @@
+                return Optional.empty();
+            }
+            int partition = Integer.parseInt(strings[0]);
+            long offset = Long.parseLong(strings[1]);
```
do we need to handle NumberFormatException here?

satishd replied:
An error will be thrown to the caller. As you suggested, it is good to catch that and return empty optional so that the caller throws an error with the details including the line. Addressed it in the latest commit.

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/RemoteLogMetadataSnapshotFile.java:
```diff
@@ -0,0 +1,239 @@
+     * @throws IOException if there4 is any error in writing the given snapshot to the file.
+     */
+    public synchronized void write(Snapshot snapshot) throws IOException {
+        File newMetadataSnapshotFile = new File(metadataStoreFile.getAbsolutePath() + ".new");
```
Our convention is to create a .tmp file.

satishd replied:
sure, updated to use the common convention.

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/RemoteLogMetadataSnapshotFile.java:
```diff
@@ -0,0 +1,239 @@
+        private final Uuid topicId;
+        private final int metadataPartition;
+        private final long metadataPartitionOffset;
+        private final Collection<RemoteLogSegmentMetadata> remoteLogMetadatas;
```
remoteLogMetadatas is weird since data is the plural form of datum.

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/RemoteLogMetadataSnapshotFile.java:
```diff
@@ -0,0 +1,239 @@
+                // Read the entry
+                ByteBuffer data = ByteBuffer.allocate(len);
+                final int read = channel.read(data);
+                if (read != len) {
```
Should we do the same check when reading the lenBuffer?

junrao commented on storage/src/main/resources/message/RemoteLogSegmentMetadataRecordSnapshot.json:
```diff
@@ -0,0 +1,92 @@
+{
+  "apiKey": 0,
+  "type": "data",
+  "name": "RemoteLogSegmentMetadataRecordSnapshot",
```
How is this class used?

satishd replied:
This is used while storing the snapshots ser/des.

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java:
```diff
@@ -66,6 +70,8 @@
     private final RemotePartitionMetadataEventHandler remotePartitionMetadataEventHandler;
     private final RemoteLogMetadataTopicPartitioner topicPartitioner;
+    private final Time time;
+
```
extra newline.

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java:
```diff
@@ -85,21 +91,67 @@
+                partitionToConsumedOffsets.put(entry.getKey(), entry.getValue());
+                consumer.seek(new TopicPartition(REMOTE_LOG_METADATA_TOPIC_NAME, entry.getKey()), entry.getValue());
+            }
+
```
extra new line.

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java:
```diff
@@ -85,21 +91,67 @@
+        Map<Integer, Long> committedOffsets = Collections.emptyMap();
+        try {
+            // load committed offset and assign them in the consumer
+            committedOffsets = committedOffsetsFile.readEntries();
```
Should we initialize committedPartitionToConsumedOffsets to the offsets in the checkpoint file?

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java:
```diff
@@ -120,6 +174,29 @@ public void run() {
         }
     }
 
+    private void syncCommittedDataAndOffsets(boolean forceSync) {
```
syncCommittedDataAndOffsets => maybeSyncCommittedDataAndOffsets ?

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java:
```diff
@@ -120,6 +174,29 @@ public void run() {
+        }
+
+        try {
+            // todo sync the snapshot file
```
Should this todo be addressed in this PR?

satishd replied:
invalid todo, this is already handled by calling remotePartitionMetadataEventHandler.syncLogMetadataDataFile(topicIdPartition, metadataPartition, offset);

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java:
```diff
@@ -120,6 +174,29 @@ public void run() {
+            }
+
+            committedOffsetsFile.writeEntries(partitionToConsumedOffsets);
+            committedPartitionToConsumedOffsets = new HashMap<>(partitionToConsumedOffsets);
```
Hmm, why do we want to reset committedPartitionToConsumedOffsets? This means if there is no new data consumed, we will still write the checkpoint file?

satishd replied:
Right, it can avoid writing if it is already synced and offset did not move further. It is addressed in the latest commit.

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java:
```diff
@@ -120,6 +174,29 @@ public void run() {
+            committedPartitionToConsumedOffsets = new HashMap<>(partitionToConsumedOffsets);
+            lastSyncedTimeMs = time.milliseconds();
+        } catch (IOException e) {
+            log.error("Error encountered while writing committed offsets to a local file", e);
```
Could we integrate this with logDirFailureChannel?

satishd replied:
For now, we will log an error if there are any issues in writing to the file. There are no direct dependencies to `logDirFailureChannel` as it is part of RLMM implementation. We can address this in PRs when remote log subsystem is added.

junrao replied:
This is an IOException from local storage. So, why do we want to wait when remote log subsystem is added? Ideally, we want to integrate it with logDirFailureChannel. If we want to do that later, it seems that we should at least throw an exception to stop the thread?

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java:
```diff
@@ -177,8 +254,9 @@ private void handleRemoteLogMetadata(RemoteLogMetadata remoteLogMetadata) {
         Set<TopicPartition> assignedMetaTopicPartitions = assignedMetaPartitionsSnapshot.stream()
-                .map(partitionNum -> new TopicPartition(REMOTE_LOG_METADATA_TOPIC_NAME, partitionNum))
-                .collect(Collectors.toSet());
+                                                                                        .map(partitionNum -> new TopicPartition(
```
The line is quite long. Could we move assignedMetaPartitionsSnapshot to a new line?

junrao commented on checkstyle/import-control.xml:
```diff
@@ -303,6 +303,12 @@
+
+      <subpackage name="remote">
+        <allow pkg="scala.collection" />
+        <allow pkg="scala.jdk" />
```
Is scala.jdk needed?

satishd replied:
Good point, this is not needed with the changes that we have.

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/RemoteLogMetadataCache.java:
```diff
@@ -161,53 +162,46 @@ public void updateRemoteLogSegmentMetadata(RemoteLogSegmentMetadataUpdate metada
-
-        doHandleSegmentStateTransitionForLeaderEpochs(existingMetadata,
+    private void handleSegmentWithCopySegmentFinishedState(RemoteLogSegmentMetadata remoteLogSegmentMetadata) {
+        doHandleSegmentStateTransitionForLeaderEpochs(remoteLogSegmentMetadata,
```
So, the current code has a bug by using existingMetadata?

satishd replied:
It is not a bug. Code is very similar but it is a minor refactoring to pass the updated RemoteLogSegmentMetadata instance.

junrao replied:
The param in doHandleSegmentStateTransitionForLeaderEpochs() is still called existingMetadata. Could we update that accordingly?

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/RemoteLogMetadataCache.java:
```diff
@@ -161,53 +162,46 @@ public void updateRemoteLogSegmentMetadata(RemoteLogSegmentMetadataUpdate metada
         // Put the entry with the updated metadata.
-        idToSegmentMetadata.put(existingMetadata.remoteLogSegmentId(),
-                existingMetadata.createWithUpdates(metadataUpdate));
+        idToSegmentMetadata.put(remoteLogSegmentMetadata.remoteLogSegmentId(), remoteLogSegmentMetadata);
```
This is an existing issue. In RemoteLogLeaderEpochState, it seems that methods like handleSegmentWithDeleteSegmentStartedState() and handleSegmentWithDeleteSegmentFinishedState() have unused input params?

satishd replied:
`doHandleSegmentStateTransitionForLeaderEpochs` takes `RemoteLogLeaderEpochState.Action` and invokes it. Each action implementer does not need all the arguments.

junrao replied:
It seems that leaderEpochEndOffset is never used below.

```
    void handleSegmentWithDeleteSegmentStartedState(Long startOffset, RemoteLogSegmentId remoteLogSegmentId,
                                                    Long leaderEpochEndOffset) {

```

It seems that startOffset and leaderEpochEndOffset are not used in the follow method.
```
    void handleSegmentWithDeleteSegmentFinishedState(long startOffset, RemoteLogSegmentId remoteLogSegmentId,
                                                     Long leaderEpochEndOffset) {

```

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/RemotePartitionMetadataStore.java:
```diff
@@ -42,21 +45,36 @@
+        if (remoteLogMetadataCache != null) {
+            remoteLogMetadataCache.addCopyInProgressSegment(remoteLogSegmentMetadata);
+        } else {
+            log.error("No partition metadata found for : " + topicIdPartition);
```
Should we throw an exception?

satishd replied:
Even if we throw an exception here, caller needs to handle these errors by logging them.  This can happen when a partition is deleted but an event for the same partition is consumed by the ConsumerTask.

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/RemotePartitionMetadataStore.java:
```diff
@@ -91,6 +109,23 @@ public void handleRemotePartitionDeleteMetadata(RemotePartitionDeleteMetadata re
+    public void syncLogMetadataDataFile(TopicIdPartition topicIdPartition,
+                                        int metadataPartition,
+                                        Long metadataPartitionOffset) throws IOException {
+        //todo-tier write partitions
```
Will the todo be addressed in this PR?

satishd replied:
This todo is not valid. It is already addressed in the existing code as `remoteLogMetadataCache.flushToFile(metadataPartition, metadataPartitionOffset);`.

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/RemotePartitionMetadataStore.java:
```diff
@@ -91,6 +109,23 @@ public void handleRemotePartitionDeleteMetadata(RemotePartitionDeleteMetadata re
+        if (partitionDeleteMetadata != null) {
+            log.info("Skipping syncing of metadata snapshot as remote partition [{}] is with state: [{}] ", topicIdPartition,
+                     partitionDeleteMetadata);
+        } else {
```
Hmm, until the partition deletion completes, it seems that we still want to persist the segment list so that we could finish deleting them from the remote store?

satishd replied:
RemotePartitionRemover may have the functionality of removing in a different way and revisit that when we have the RemotePartitionRemover functionality.
For now, the cache is stored as part of the topic partition log directory and it will be deleted when the respective partition is deleted locally by a broker.

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/TopicBasedRemoteLogMetadataManager.java:
```diff
@@ -81,12 +83,21 @@
     private final Set<TopicIdPartition> pendingAssignPartitions = Collections.synchronizedSet(new HashSet<>());
     private volatile boolean initializationFailed;
 
+    public TopicBasedRemoteLogMetadataManager() {
```
This seems unused?

satishd replied:
This will be used when the remote log subsystem is added and it will be used as the default RLMM if it is not configured by a user.

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/RemoteLogMetadataSnapshotFile.java:
```diff
@@ -0,0 +1,239 @@
+            return metadataPartition;
+        }
+
+        public long metadataPartitionOffset() {
```
This seems unused?

satishd replied:
This is used to store it in the snapshot file.

junrao replied:
Yes, but it seems that we never read it?

satishd replied:
This is added more for debugging purposes. This was suggested while the KIP was reviewed.

junrao replied:
Since we flush the consumer offset file after flushing the snapshot file, it's possible for messages to be replayed on broker restart. Should we use metadataPartitionOffset to avoid duplicated messages being reapplied to the remote log segment metadata cache?

satishd replied:
Yes, that can be done. I will address it in a followup PR.

junrao commented on storage/src/main/resources/message/RemotePartitionDleteMetadataSnapshot.json:
```diff
@@ -0,0 +1,42 @@
+{
+  "apiKey": 1,
+  "type": "data",
+  "name": "RemotePartitionDeleteMetadataSnapshot",
```
How is this class used?

satishd replied:
Removed for now as it is not needed now. We will revisit this later if needed.

junrao reviewed (commented):
@satishd : Thanks for the PR. Made a pass of non-testing files. A few comments below.

satishd committed:
Addressed review comments

satishd commented:
@junrao : Thanks for the review comments. Sorry about the wrong branch merge which caused the confusion here. Addressed the comments in the latest commit.

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/RemoteLogMetadataSnapshotFile.java:
```diff
@@ -0,0 +1,252 @@
+            if (!(o instanceof Snapshot)) return false;
+            Snapshot snapshot = (Snapshot) o;
+            return version == snapshot.version && metadataPartition == snapshot.metadataPartition && metadataPartitionOffset == snapshot.metadataPartitionOffset && Objects
+                    .equals(topicId, snapshot.topicId);
```
Should we include remoteLogSegmentMetadataSnapshots too?

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/RemoteLogMetadataSnapshotFile.java:
```diff
@@ -0,0 +1,252 @@
+                    "version=" + version +
+                    ", topicId=" + topicId +
+                    ", metadataPartition=" + metadataPartition +
+                    ", metadataPartitionOffset=" + metadataPartitionOffset +
```
Should we include some metadata (e.g. size) of remoteLogSegmentMetadataSnapshots?

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/serialization/RemoteLogSegmentMetadataSnapshotRecordTransform.java:
```diff
@@ -0,0 +1,98 @@
+import java.util.Map;
+import java.util.stream.Collectors;
+
+public class RemoteLogSegmentMetadataSnapshotRecordTransform implements RemoteLogMetadataTransform<RemoteLogSegmentMetadata> {
```
This class seems never used?

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java:
```diff
@@ -85,21 +90,68 @@
     @Override
     public void run() {
         log.info("Started Consumer task thread.");
+        lastSyncedTimeMs = time.milliseconds();
```
Below, maybeWaitForPartitionsAssignment() may add new assignedMetaPartitions. We need to seek those partitions from the beginning.

satishd replied:
I guess the consumer would fetch from the earliest offset as there is no state associated with in the consumer for those partitions as `auto.offset.reset` is set as `earliest`.

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java:
```diff
@@ -120,6 +174,35 @@ public void run() {
+
+        try {
+            HashMap<Integer, Long> syncedPartitionToConsumedOffsets = new HashMap<>();
+            for (TopicIdPartition topicIdPartition : assignedTopicPartitions) {
```
assignedTopicPartitions could be updated concurrently and we are accessing it without lock protection here.

satishd replied:
Good point. Added the locking and updated it with a comment.

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java:
```diff
@@ -120,6 +174,35 @@ public void run() {
+                }
+            }
+
+            committedOffsetsFile.writeEntries(partitionToConsumedOffsets);
```
Some of the metadataPartitions could have been removed. Should we write syncedPartitionToConsumedOffsets instead? Should we also set partitionToConsumedOffsets to syncedPartitionToConsumedOffsets?

satishd replied:
We needed partitionToConsumedOffsets in the earlier check and we do not really need syncedPartitionToConsumedOffsets here. Updated to use partitionToConsumedOffsets for writing and setting committedPartitionToConsumedOffsets as partitionToConsumedOffsets.

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/serialization/RemoteLogMetadataSerde.java:
```diff
@@ -39,6 +41,7 @@
     private static final short REMOTE_LOG_SEGMENT_METADATA_API_KEY = new RemoteLogSegmentMetadataRecord().apiKey();
     private static final short REMOTE_LOG_SEGMENT_METADATA_UPDATE_API_KEY = new RemoteLogSegmentMetadataUpdateRecord().apiKey();
     private static final short REMOTE_PARTITION_DELETE_API_KEY = new RemotePartitionDeleteMetadataRecord().apiKey();
+    private static final short REMOTE_LOG_SEGMENT_METADATA_SNAPSHOT_API_KEY = new RemoteLogSegmentMetadataSnapshotRecord().apiKey();
```
REMOTE_LOG_SEGMENT_METADATA_SNAPSHOT is not in internal remote log metadata topic and the comment of the class says this is for messages in remote log metadata topic. So, do we need this change?

satishd replied:
There are two ways to implement it. 
- Have a separate BytesApiMessageSerde for this message only and write the supporting classes which will be similar to RemoteLogMetadataSerde and RemoteLogSegmentMetadataSnapshotTransform.
- This is one more ApiMessage about representing remote log metadata. Add to the existing RemoteLogMetadatSerde which has the framework to add one more api message. 

I choose the latter for simplicity. We can update the javadoc to describe that RemoteLogMetadatSerde includes serde for all the APIMessage defined for remote log metadata including RemoteLogSegmentMetadataSnapshot. It can be used as serde for the topic as it supports all the messages stored in the remote log metadata topic.

junrao replied:
Hmm, since the remote log snapshot file has a file level header, having the frame header at record level seems redundant. But for simplicity, we probably could just write the framed record to the snapshot file. Could we update the comment accordingly?

satishd replied:
I do not see any redundancy here.
File level header contains 
` <version:short><metadata-partition:int><metadata-partition-offset:long>`
This is not repeated in each entry. Pl let me know if I am missing anything here.

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/RemoteLogMetadataSnapshotFile.java:
```diff
@@ -0,0 +1,252 @@
+
+    public static final String COMMITTED_LOG_METADATA_SNAPSHOT_FILE_NAME = "remote_log_snapshot";
+
+    // header: <version:short><topicId:2 longs><metadata-partition:int><metadata-partition-offset:long>
```
Why do we need topicId since this file is already inside a topic partition?

satishd replied:
This was suggested in KIP review to add this as it will be useful for debugging. To make sure that, what is stored in the file and the received topic id should be the same.

junrao replied:
The thing is that we already store topicId in a partitionMetadata file (https://cwiki.apache.org/confluence/display/KAFKA/KIP-516%3A+Topic+Identifiers#KIP516:TopicIdentifiers-PartitionMetadatafile). Storing the same info in another file seems to add confusion.

satishd replied:
I removed it as suggested.

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/RemoteLogMetadataCache.java:
```diff
@@ -161,53 +161,46 @@ public void updateRemoteLogSegmentMetadata(RemoteLogSegmentMetadataUpdate metada
-        log.debug("Adding remote log segment metadata to leader epoch mappings with update: [{}]", metadataUpdate);
-
-        doHandleSegmentStateTransitionForLeaderEpochs(existingMetadata,
+    protected final void handleSegmentWithCopySegmentFinishedState(RemoteLogSegmentMetadata remoteLogSegmentMetadata) {
```
RemoteLogLeaderEpochState::handleSegmentWithCopySegmentFinishedState() has the logic to move the existing segment with the same offset in offsetToId to unreferencedSegmentIds, which seems sensitive to ordering. However, the snapshot are stored in an arbitrary order. Is that an issue?

satishd replied:
Good catch. We should maintain the ordering while storing the segments.

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/RemoteLogSegmentMetadataSnapshot.java:
```diff
@@ -0,0 +1,200 @@
+                ", endOffset=" + endOffset +
+                ", maxTimestampMs=" + maxTimestampMs +
+                ", segmentSizeInBytes=" + segmentSizeInBytes +
+                ", state=" + state +
```
Should toString() include segmentLeaderEpochs too?

junrao reviewed (commented):
@satishd : Thanks for the updated PR. A few more comments below.

ccding commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerManager.java:
```diff
@@ -49,21 +52,23 @@
         KafkaConsumer<byte[], byte[]> consumer = new KafkaConsumer<>(rlmmConfig.consumerProperties());
-        consumerTask = new ConsumerTask(consumer, remotePartitionMetadataEventHandler, rlmmTopicPartitioner);
+        Path committedOffsetsPath = new File(rlmmConfig.logDir(), COMMITTED_OFFSETS_FILE_NAME).toPath();
+        consumerTask = new ConsumerTask(consumer, remotePartitionMetadataEventHandler, topicPartitioner, committedOffsetsPath, time, 60_000L);
```
do we plan to make `60_000L` a variable/configurable in the future? like other intervals in KafkaConfig.

satishd replied:
Right, this will be added in a followup PR.

ccding commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java:
```diff
@@ -85,21 +90,68 @@
+            log.error("Encountered error while building committed offsets from the file", e);
+        }
+
+        final Set<Map.Entry<Integer, Long>> entries = committedOffsets.entrySet();
```
out of curiosity: what is the purpose of converting a map to a set of map entries?

satishd replied:
We want to go through the entries instead of calling get(key) with all the keys avoiding the lookups.

satishd committed:
Addressed review comments.
Added a test for loading the snapshot and check whethr segments are loaded in the expected order.

satishd commented:
Thanks @junrao for the review. Addressed them with the latest commit and comments.

satishd commented:
Thanks @ccding for the review, addressed with inline replies.

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java:
```diff
@@ -85,32 +90,78 @@
+
+            // Seek to the committed offsets
+            for (Map.Entry<Integer, Long> entry : committedOffsets.entrySet()) {
+                partitionToConsumedOffsets.put(entry.getKey(), entry.getValue());
```
If the metadataTopicPartitions changes, should we remove unneeded partitions from partitionToConsumedOffsets?

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java:
```diff
@@ -85,32 +90,78 @@
     // Map of remote log metadata topic partition to consumed offsets.
     private final Map<Integer, Long> partitionToConsumedOffsets = new ConcurrentHashMap<>();
 
+    private Map<Integer, Long> committedPartitionToConsumedOffsets = Collections.emptyMap();
```
To be consistent with the naming of partitionToConsumedOffsets, would it be better to name this partitionToCommittedOffsets?

satishd replied:
This logic is slightly changed in the latest commit for easy to reason about. Below are the updated maps to track. 

```
    // Map of remote log metadata topic partition to consumed offsets. Received consumer records
    // may or may not have been processed based on the assigned topic partitions.
    private final Map<Integer, Long> partitionToConsumedOffsets = new ConcurrentHashMap<>();

    // Map of remote log metadata topic partition to processed offsets. Received consumer record is
    // processed as the remote log metadata record's topic partition exists in assigned topic partitions.
    private final Map<Integer, Long> partitionToProcessedOffsets = new ConcurrentHashMap<>();

    // Map of remote log metadata topic partition to processed offsets that were synced in committedOffsetsFile.
    private Map<Integer, Long> lastSyncedPartitionToConsumedOffsets = Collections.emptyMap();
```

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java:
```diff
@@ -120,6 +171,37 @@ public void run() {
+                for (TopicIdPartition topicIdPartition : assignedTopicPartitions) {
+                    int metadataPartition = topicPartitioner.metadataPartition(topicIdPartition);
+                    Long offset = partitionToConsumedOffsets.get(metadataPartition);
+                    if (offset != null && !offset.equals(committedPartitionToConsumedOffsets.get(metadataPartition))) {
```
This means that if there is no change to a remoteLogMetadataCache, but there is new record for other partitions in the same metadataPartition, we still need to flush remoteLogMetadataCache.

satishd replied:
This is a known limitation that I want to address in the future. One possible way I was thinking is to keep track of the last processed event’s offset to be maintained in the in-memory store and the last offset written into the snapshot file. If both these entries are the same then we will skip taking a new snapshot and committing it.

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/RemoteLogMetadataSnapshotFile.java:
```diff
@@ -0,0 +1,253 @@
+    public static final String COMMITTED_LOG_METADATA_SNAPSHOT_FILE_NAME = "remote_log_snapshot";
+
+    // header: <version:short><topicId:2 longs><metadata-partition:int><metadata-partition-offset:long>
+    // size: 2 + (8+8) + 4 + 8 = 30
```
Could we describe the format of the rest of the content too?

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/RemoteLogMetadataSnapshotFile.java:
```diff
@@ -0,0 +1,253 @@
+            }
+        }
+
+        Utils.atomicMoveWithFallback(newMetadataSnapshotFile.toPath(), metadataStoreFile.toPath());
```
Should we force the channel at the end?

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java:
```diff
@@ -240,6 +323,11 @@ public void close() {
                 closing = true;
                 consumer.wakeup();
+                try {
+                    maybeSyncCommittedDataAndOffsets(true);
```
Should we wait until the consumer task completes before writing the checkpoint file? Otherwise, we could be reading the in-memory state while it's being updated.

satishd replied:
Another way is to call `maybeSyncCommittedDataAndOffsets` in finally block of run method before `closeConsumer()`. This will avoid running in multiple threads and the state that is read is consistent.

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/RemoteLogMetadataCache.java:
```diff
@@ -161,77 +161,73 @@ public void updateRemoteLogSegmentMetadata(RemoteLogSegmentMetadataUpdate metada
         // Put the entry with the updated metadata.
-        idToSegmentMetadata.put(existingMetadata.remoteLogSegmentId(),
-                existingMetadata.createWithUpdates(metadataUpdate));
+        idToSegmentMetadata.put(remoteLogSegmentMetadata.remoteLogSegmentId(), remoteLogSegmentMetadata);
```
This is an existing issue. When removing old segments, should we remove entries from leaderEpochEntries too?

satishd replied:
`remoteLogLeaderEpochState.handleSegmentWithDeleteSegmentStartedState(startOffset, segmentId)` call in the earlier statement removes the entries. Let me know if I am missing something here.

junrao replied:
Got it. You have a comment that leaderEpochEntries will be removed in a later jira.

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/RemoteLogSegmentMetadataSnapshot.java:
```diff
@@ -0,0 +1,201 @@
+import java.util.Objects;
+import java.util.TreeMap;
+
+public class RemoteLogSegmentMetadataSnapshot extends RemoteLogMetadata {
```
Could we add a comment for this class?

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/RemoteLogSegmentMetadataSnapshot.java:
```diff
@@ -0,0 +1,201 @@
+public class RemoteLogSegmentMetadataSnapshot extends RemoteLogMetadata {
+
+    /**
+     * Universally unique remote log segment id.
```
I am a bit confused. This class seems to be the same as RemoteLogSegmentMetadata?

satishd replied:
Updated with the javadoc.

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/RemotePartitionMetadataStore.java:
```diff
@@ -42,21 +45,36 @@
+        if (remoteLogMetadataCache != null) {
+            remoteLogMetadataCache.addCopyInProgressSegment(remoteLogSegmentMetadata);
+        } else {
+            log.warn("No partition metadata found for : " + topicIdPartition);
```
If this is unexpected, should we throw an IllegalStateException?

junrao reviewed (commented):
@satishd : Thanks for the updated PR. A few more comments on non-testing files.

satishd committed:
Addressed review comments.
 - Removed topicId from RemoteLogMetadataSnapshotFile.Snapshot
 - ConsumerTask handles processed or committed offsets better.
 - Minor cleanup for better readability.

satishd commented:
Thanks @junrao for the review. Addressed them with the replies and with the latest commit.

junrao commented on storage/src/main/resources/message/RemoteLogSegmentMetadataSnapshot.json:
```diff
@@ -0,0 +1,92 @@
+{
+  "apiKey": 3,
+  "type": "metadata",
+  "name": "RemoteLogSegmentMetadataSnapshotRecord",
```
To be consistent, should the file name be RemoteLogSegmentMetadataSnapshotRecord.json?

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java:
```diff
@@ -82,44 +87,148 @@
+        synchronized (assignPartitionsLock) {
+            if (assignedTopicPartitions.contains(remoteLogMetadata.topicIdPartition())) {
+                remotePartitionMetadataEventHandler.handleRemoteLogMetadata(remoteLogMetadata);
+                partitionToProcessedOffsets.put(record.partition(), record.offset());
```
Maintaining partitionToProcessedOffsets here seems a bit weird since it's not exactly the processed offset for a particular partition. Could we pass along the record offset to remotePartitionMetadataEventHandler.handleRemoteLogMetadata() and let RemotePartitionMetadataStore remember the last processed offset?

satishd replied:
As I mentioned in my earlier [comment](https://github.com/apache/kafka/pull/11058#discussion_r719295072), we will be passing offset and storing it as part of inmemory and snapshots. This will be done in a followup PR. With that, we do not need to maintain here.

junrao replied:
Since the processOffset in the checkpoint file is not really being used for now and we plan to get rid of it in the future, could we just use consumed offset for now?

junrao commented on storage/src/test/java/org/apache/kafka/server/log/remote/metadata/storage/RemoteLogMetadataSnapshotFileTest.java:
```diff
@@ -0,0 +1,82 @@
+        RemoteLogMetadataSnapshotFile snapshotFile = new RemoteLogMetadataSnapshotFile(metadataStoreDir.toPath());
+
+        // There should be an empty snapshot as nothing is written into it.
+        Assertions.assertFalse(snapshotFile.read().isPresent());
```
It's easier to understand if we assertTrue on isEmpty.

satishd replied:
[Optional](https://docs.oracle.com/javase/8/docs/api/java/util/Optional.html) does not have `isEmpty()` method in JDK 8, that's why using `isPresent ()` and the respective check. `isEmpty()` is available since JDK 11.

junrao commented on storage/src/test/java/org/apache/kafka/server/log/remote/metadata/storage/TopicBasedRemoteLogMetadataManagerRestartTest.java:
```diff
@@ -0,0 +1,156 @@
+    public void testRLMMAPIsAfterRestart() throws Exception {
+        // Create topics.
+        String leaderTopic = "new-leader";
+        HashMap<Object, Seq<Object>> assignedLeaderTopicReplicas = new HashMap<>();
```
Should we use `HashMap<Integer, Seq<Integer>>`?

satishd replied:
I wrote like that initially and hit the below Scala quirk before it is changed to `HashMap<Object, Seq<Object>>`, 
```
/Users/satishd/repos/kafka/storage/src/test/java/org/apache/kafka/server/log/remote/metadata/storage/TopicBasedRemoteLogMetadataManagerRestartTest.java:97: error: incompatible types: inference variable A has incompatible equality constraints Object,Integer
        remoteLogMetadataManagerHarness.createTopic(leaderTopic, JavaConverters.mapAsScalaMap(assignedLeaderTopicReplicas));
                                       ^
  where A,B are type-variables:
    A extends Object declared in method <A,B>mapAsScalaMap(java.util.Map<A,B>)
    B extends Object declared in method <A,B>mapAsScalaMap(java.util.Map<A,B>)
```

ccding replied:
```        HashMap<Integer, Seq<Integer>> assignedLeaderTopicReplicas = new HashMap<>();
        List<Integer> leaderTopicReplicas = new ArrayList<>();
        // Set broker id 0 as the first entry which is taken as the leader.
        leaderTopicReplicas.add(0);
        leaderTopicReplicas.add(1);
        leaderTopicReplicas.add(2);
        assignedLeaderTopicReplicas.put(0, JavaConverters.asScalaBuffer(leaderTopicReplicas));
        remoteLogMetadataManagerHarness.createTopic(leaderTopic,
            (scala.collection.Map<Object, Seq<Object>>) assignedLeaderTopicReplicas);
```
It appears this would work.

junrao commented on storage/src/test/java/org/apache/kafka/server/log/remote/metadata/storage/TopicBasedRemoteLogMetadataManagerRestartTest.java:
```diff
@@ -0,0 +1,156 @@
+        // Create topics.
+        String leaderTopic = "new-leader";
+        HashMap<Object, Seq<Object>> assignedLeaderTopicReplicas = new HashMap<>();
+        List<Object> leaderTopicReplicas = new ArrayList<>();
```
Should we use` List<Integer>`?

satishd replied:
Similar Scala quirk as I mentioned in my earlier comment.

junrao reviewed (commented):
@satishd : Thanks for the updated PR. Made a pass of all files. A few more comments.

Also, should we add a test on the checkpointed consumer offset?

satishd committed:
Addressed review comments.
  - Renamed RemoteLogSegmentMetadataSnapshot.json RemoteLogSegmentMetadataSnapshotRecord.json
  - Added a scenario to check committed offsets when TopicBasedRemoteLogMetadataManager is closed and make sure it is reloaded with the earleir committed offsets.

satishd committed:
Empty commit to kick off jenkins build.

satishd commented:
Thanks @junrao for the review comments. Addressed them with the latest commit and replies. 

Added a scenario to verify checkpointed offsets in the [test](https://github.com/apache/kafka/pull/11058/commits/554df8c5e58f5dc14b5d1a3476f011184116a088#diff-8c57d1a1451531841bccd4de7f38b838cfa8444e0257c5097480e92e3e0fe72bR145).

kowshik commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/CommittedOffsetsFile.java:
```diff
@@ -0,0 +1,83 @@
+
+        @Override
+        public Optional<Map.Entry<Integer, Long>> fromString(String line) {
+            String[] strings = MINIMUM_ONE_WHITESPACE.split(line);
```
Would `line.split(SEPARATOR)` be simpler?

satishd replied:
I am not sure how would that be simpler. `String.split(regex)` uses `Pattern.compile(regex).split()` for this usecase and it needs to compile regex for each invocation of fromString() method. 
So, I decided to use `MINIMUM_ONE_WHITESPACE.split(line)` as `Pattern.compile()` is done only once, which is more efective

kowshik replied:
It seems performance optimization is not a concern here, since deserialization of committed offsets file from disk happens only during initialization.

kowshik commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/RemoteLogMetadataSnapshotFile.java:
```diff
@@ -0,0 +1,253 @@
+    // File format:
+    // <header>[<entry>...]
+    // header: <version:short><metadata-partition:int><metadata-partition-offset:long>
+    // entry: <entry-length><entry-bytes>
```
Could we additionally store the # of entries in the header, so that during deserialization from disk we can pre-allocate the size of the result set?

kowshik commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/RemoteLogMetadataSnapshotFile.java:
```diff
@@ -0,0 +1,253 @@
+
+        // Create an empty file if it does not exist.
+        try {
+            boolean newFileCreated = metadataStoreFile.createNewFile();
```
Do we need to create the file here? It seems in L126 the call to `Utils.atomicMoveWithFallback` will create the target file if it did not exist earlier.

satishd replied:
This is to make fail fast if there is any issue in creating a file.

kowshik commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/RemoteLogMetadataSnapshotFile.java:
```diff
@@ -0,0 +1,253 @@
+    public synchronized Optional<Snapshot> read() throws IOException {
+
+        // Checking for empty files.
+        if (metadataStoreFile.length() == 0) {
```
If we didn't create an empty file in the constructor, would we be needing this check for emptiness?

satishd replied:
No, it is not needed in that case.

kowshik commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/RemoteLogMetadataSnapshotFile.java:
```diff
@@ -0,0 +1,253 @@
+                        int metadataPartition,
+                        long metadataPartitionOffset,
+                        Collection<RemoteLogSegmentMetadataSnapshot> remoteLogSegmentMetadataSnapshots) {
+            this.version = version;
```
Hmm, what happens if `version` doesn't match `CURRENT_VERSION`?

satishd replied:
There is only one version for now. Added a guard against checking with the expected version.

junrao commented on storage/src/test/java/org/apache/kafka/server/log/remote/metadata/storage/TopicBasedRemoteLogMetadataManagerRestartTest.java:
```diff
@@ -0,0 +1,180 @@
+
+        int metadataPartition1 = topicBasedRlmm().metadataPartition(leaderTopicIdPartition);
+        int metadataPartition2 = topicBasedRlmm().metadataPartition(followerTopicIdPartition);
+        Optional<Long> receivedOffsetForPartition1 = topicBasedRlmm().receivedOffsetForPartition(metadataPartition1);
```
Since RLMM.initializeResources() is called in a separate thread, there is no guarantee that the ConsumerManager has been set up at this point. Should we put the check under waitUntil()?

satishd replied:
`remoteLogMetadataManagerHarness.initialize(Collections.emptySet(), true)` is already called in `setup()` method which waits until RLMM is initialized as you can see [here](https://github.com/apache/kafka/blob/trunk/storage/src/test/java/org/apache/kafka/server/log/remote/metadata/storage/TopicBasedRemoteLogMetadataManagerHarness.java#L73).

junrao commented on storage/src/test/java/org/apache/kafka/server/log/remote/metadata/storage/TopicBasedRemoteLogMetadataManagerRestartTest.java:
```diff
@@ -0,0 +1,180 @@
+        Assertions.assertTrue(receivedOffsetForPartition2.isPresent());
+
+        // Make sure these offsets are at least 0.
+        Assertions.assertTrue(receivedOffsetForPartition1.get() >= 0);
```
Hmm, are we testing anything useful here since offset is always >= 0?

satishd replied:
If there is no entry for this partition, it will be empty and we already have a check for that.  This offset represents messages until this offset has been received by RLMM.  So, if the received offset is 0 means it received message until 0 offset. 
This check is to guard against any future implementation like having a default entry for all the registered metadata partitions with a value < 0.

junrao reviewed (commented):
@satishd : Thanks for the updated PR. A couple of more comments.

kowshik commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/CommittedOffsetsFile.java:
```diff
@@ -0,0 +1,83 @@
+    public synchronized Map<Integer, Long> readEntries() throws IOException {
+        List<Map.Entry<Integer, Long>> entries = checkpointFile.read();
+        Map<Integer, Long> partitionToOffsets = new HashMap<>(entries.size());
+        for (Map.Entry<Integer, Long> entry : entries) {
```
Should we guard against a duplicate entry on disk (can happen only during corruption)?

satishd replied:
Good point. I will add a check against that.

kowshik commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java:
```diff
@@ -82,44 +87,148 @@
+            committedOffsets = committedOffsetsFile.readEntries();
+        } catch (IOException e) {
+            // Ignore the error and consumer consumes from the earliest offset.
+            log.error("Encountered error while building committed offsets from the file", e);
```
Should the error message mention that the consumer is falling back to replay from earliest offset?

kowshik commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java:
```diff
@@ -82,44 +87,148 @@
     // User topic partitions that this broker is a leader/follower for.
     private Set<TopicIdPartition> assignedTopicPartitions = Collections.emptySet();
```
There is `assignPartitionsLock` lock defined above. It would be useful to document above it what attributes/behavior it guards.

kowshik commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java:
```diff
@@ -82,44 +87,148 @@
+                    processConsumerRecord(record);
                 }
+
+                maybeSyncCommittedDataAndOffsets(false);
```
Hmm, is this method the only place where we would sync the data to disk? I thought we would want to sync to disk periodically as well.

satishd replied:
We wanted to do this in the same thread for now. We will do it in a separate thread if it is really needed.

kowshik commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java:
```diff
@@ -82,44 +87,148 @@
+            }
+
+            lastSyncedTimeMs = time.milliseconds();
+        } catch (IOException e) {
```
Should we eventually shutdown the broker due to the IOException?
IIUC, currently the exception gets logged and only the consumer thread dies.

cc @junrao

satishd replied:
Good point. As discussed in syncup call, we will address it in a followup PR. Filed [KAFKA-13355](  https://issues.apache.org/jira/browse/KAFKA-13355) to track this issue.

kowshik reviewed (commented):
@satishd Thanks for the PR! Made a pass on non-testing files.

satishd commented:
Thanks @junrao for the comments, added in-line replies.

satishd committed:
Addressed review comments
 - Added entries size in RemoteLogMetadataSnapshotFile.
 - Added defensive check while reading duplicate entries in CommittedOffsetsFile.
 - Other minor javadoc updates and cleanup..

junrao reviewed (commented):
@satishd : Thanks for the updated PR. Just one more comment below.

satishd committed:
Addressed review comments.
 - Removed partitionToProcessedOffsets in ConsumerTask.

satishd commented:
Thanks @junrao for your comment, Addressed with the latest commit.

kowshik reviewed (approved):
@satishd Thanks for the updated PR. LGTM.

junrao reviewed (approved):
@satishd : Thanks for the updated PR. LGTM
